server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java

```diff
@@ -287,14 +287,19 @@ private void writeValue(Object value, MediaType requested, MediaType configuredM
       if (returnBody) responseBuilder.entity(value);
    }
 
-   private <V> String calcETAG(V value) {
-      return String.valueOf(hashFunc.hash(value));
+   private String calcETAG(Metadata metadata) {
+      LongEntryVersion entryVersion = (LongEntryVersion) metadata.version();
+      if (entryVersion != null) {
```

Comment:
Under what circumstances the entry does not have a version?

Reply:
When the entry is not inserted via Hot Rod or Rest

Reply:
Hmm, I see that it is not enforced, it seems that server is adding the version info. I am thinking about adding VersionInterceptor in server all the time, or enforcing that somewhere else... otherwise the conditional ops won't work correctly.

Reply:
```adding VersionInterceptor in server all the time,```

Would that cover the case when someone executes a script or a task  in the server getting hold of a cache directly and writing an entry?

Reply:
`VersionInterceptor` doesn't really add versions to new entries @rvansa. And there's still the possibility of embedded nodes running in the same cluster with server nodes... I'd rather force embedded to use versions all the time as well, if the perf hit isn't too big.

Reply:
@danberindei Both generators should be deleted :)

Reply:
@danberindei Scattered cache relies on topology ids in the entry version (though it does not use the version generator).